packages/react-dom-bindings/src/server/ReactDOMServerExternalRuntime.js

```diff
@@ -13,13 +13,65 @@ import {
   completeSegment,
 } from './fizz-instruction-set/ReactDOMFizzInstructionSet';
 
-// Intentionally does nothing. Implementation will be added in future PR.
-// eslint-disable-next-line no-unused-vars
-const observer = new MutationObserver(mutations => {
-  // These are only called so I can check what the module output looks like. The
-  // code is unreachable.
-  clientRenderBoundary();
-  completeBoundaryWithStyles();
-  completeBoundary();
-  completeSegment();
-});
+// This runtime may be sent to the client multiple times (if FizzServer.render
+//  is called more than once). Here, we check whether the mutation observer
```

Comment:
> Even if we accidentally got two, it seems like it should be able to work because we remove the instruction set right now, so we can check if we've consumed it.

That makes sense! A few questions I had:
- Is the goal to optimize for code size (and the expected case of only sending this script once)? In the case of multiple mutation observers being installed, would the perf tradeoff (e.g. executing the callback multiple times) be negligible? 
- (Just to check that I'm understanding) is this related to your other comment about `parentNode` (e.g. checking that it is non-null, or not yet processed)

> Given that we need to know where to properly emit the runtime and that is typically something that would go into the <head>. If you do render separate inner roots?

If I'm interpreting correctly, does this mean that we should be emitting this script tag in `head` instead of `body`? (Sorry, could be completely misunderstanding here)

Reply:
>  If we're emitting the full content, we don't emit the runtime script since that section doesn't need it.

Ah that makes sense, this must be what you're referring to for the test cases as well.

Reading through `react-server/ReactFizzServer`, would we implement this by conditionally writing the external runtime, based on `request.allPendingTasks / clientRenderedBoundaries / completedBoundaries / partialBoundaries`? (after or around [here](https://github.com/facebook/react/blob/093d4d404a4102fb8e3a5f56914e10267246495e/packages/react-server/src/ReactFizzServer.js#L2296))

Reply:
> Is the goal to optimize for code size (and the expected case of only sending this script once)? In the case of multiple mutation observers being installed, would the perf tradeoff (e.g. executing the callback multiple times) be negligible?

The goal is to optimize for the case where everything is configured correctly. If that affects perf negatively for a misconfigured runtime, which is already negatively affected, then that's ok because the fix is to fix the configuration.

Configuring your own SSR without using a framework is already an advanced topic. Configuring your own streaming SSR even more so. Configuring that with an external runtime is like three levels deep advanced topics. So we're not trying to optimize for n00b mistakes.

The important thing here is just that there is *some* way to configure it correctly at optimal size/perf.

Reply:
I think we'll need to treat this external script as a "resource" and only include that resource as part of the shell if we know that it's needed (there are still pending boundaries that hasn't completed or hasn't been inlined).

If you know you'll likely to need it even earlier you can call `preinit(script)` to get it into the preamble.

@gnoff Can you help with how this should be implemented?

Reply:
@mofeiZ what I'd reccomend here is to extend `writeInitialResources` to take an argument for whether the shell will flush all segments. essentially this is `request.allPendingTasks === 0`. Or you could invert it as `willFlushLateSegments` with `request.allPendingTasks > 0`.

Then if we need the runtime (based on config/flags) we call `preinit` on that script before doing the actual flushing. We may want to just expose a modified `preinit` that accepts the resources argument directly since this otherwise needs to be prepared and cleaned up similar to what happens in `retryTask`.

The reason we would use the preinit path is so that if someone already preinit'd it, we would noop rather than be emitted twice

Reply:
So if it turns out we can emit the entire response as the shell, no runtime need be transmitted, but we also don't have to wait for a late segment to get written for the script to start download when there will be late segments

Reply:
That makes sense, thanks for the clarification! I updated the PR to include this.

My only concern - if I'm reading the [async spec](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-async) correctly, it looks like the external runtime may execute before the `<body>` start tag is parsed (unlikely, but possible if it was locally cached). If this is possible, `document.body` could be null, meaning that the mutation observer could fail to attach.

@gnoff Is this possible? If so, any advice on how to address this? I searched around, but I think `onload` and similar events would fire too late (after parsing the document)

Reply:
@mofeiZ yeah if you preinit the runtime it very likely could run before the body is created. that said, browsers do some funky stuff when parsing HTML and it's possible there will already be a body and then when the streamed in body arrives it just adopts the attributes and inserts its children in the existing body. I've seen some strange stuff with parsing rules. That said, is there a reason you don't observer the Document itself or `document.documentElement`?. the documentElement will always be there becasue we never stream anything before the `<html>` open tag. This seems like a simple workaround to avoid a potentially rare but problematic race with the body tag

Reply:
@gnoff  Gotcha, that confirms what I was worried about as well.

The mutation observer is currently observing changes to direct children of the body. We could change it to call observe on `document.documentElement` with `subtree=true`, but the callback would trigger many more times (unnecessarily, per Seb's previous comment).

Maybe something like this would work:
```
if (document.body) {
  mutationObserver.observe((document.body), { childList: true, });
} else {
  // body doesn't yet exist, wait for it to be added
  const documentObserver = new MutationObserver(() => {
    for (let i = 0; i < mutations.length; i++) {
      const addedNodes = mutations[i].addedNodes;
      for (let j = 0; j < addedNodes.length; j++) {
        if (addedNodes.item(j).tagName === 'BODY') {
           // start listening for SSR instructions
           mutationObserver.observe((document.body), { childList: true, });
           documentObserver.disconnect()
         }
  }}});
  documentObserver.observe(document.documentElement, { childList: true});
}
```

Reply:
Yeah I think this is probably the way to go